src/decoder.rs

```diff
@@ -1,128 +1,126 @@
-use crate::encoder;
+use crate::NomResult;
 use crate::errors::OscError;
 use crate::types::{
-    OscArray, OscBundle, OscColor, OscMessage, OscMidiMessage, OscPacket, OscTime, OscType, Result,
+    OscArray,
+    OscBundle,
+    OscColor,
+    OscMessage,
+    OscMidiMessage,
+    OscPacket,
+    OscTime,
+    OscType,
 };
 
-use std::io::{BufRead, Read};
-use std::{char, io};
+use crate::alloc::{
+    string::{ String, ToString },
+    vec::Vec,
+};
 
-use byteorder::{BigEndian, ReadBytesExt};
+use nom::Offset;
+use nom::sequence::terminated;
+use nom::bytes::complete::{take, take_till};
+use nom::combinator::{map, map_parser};
+use nom::multi::many0;
+use nom::number::complete::{be_f32, be_f64, be_i32, be_i64, be_u32};
+use nom::{
+    IResult,
+    combinator::map_res,
+    sequence::tuple
+};
 
 /// Common MTU size for ethernet
 pub const MTU: usize = 1536;
 
 /// Takes an byte slice as argument and returns an
 /// OSC packet on success or an `OscError` if the slice
 /// does not contain a valid OSC message.
-pub fn decode(msg: &[u8]) -> Result<OscPacket> {
-    if msg.is_empty() {
-        return Err(OscError::BadPacket("Empty packet."));
+pub fn decode(msg: &[u8]) -> NomResult<OscPacket> {
+    decode_packet(msg, msg).map(|(_, packet)| packet)
+}
+
+fn decode_packet<'a>(input: &'a [u8], original_input: &'a[u8]) -> IResult<&'a [u8], OscPacket, OscError> {
+    if input.is_empty() {
+        return Err(nom::Err::Error(OscError::BadPacket("Empty packet.")));
     }
 
-    match msg[0] as char {
-        '/' => decode_message(msg),
-        '#' => decode_bundle(msg),
-        _ => Err(OscError::BadPacket("Unknown message format.")),
+    match input[0] as char {
+        '/' => decode_message(input, original_input),
+        '#' => decode_bundle(input, original_input),
+        _ => Err(nom::Err::Error(OscError::BadPacket("Unknown message format."))),
     }
```

Comment:
oh yeah, perhaps we want to add a `fn decode_tcp` or `fn decode_framed` for this then? That doesn't sound too bad to add at all - I'm imagining we don't even need to do anything other then skip that frame length byte since like you say the parser will just return the remaining un-parsed slice. So then a user just calls it in a loop until it stops returning messages/bundles.

Reply:
TCP parsing is now supported - I've added `decode_tcp` as well as `decode_tcp_vec` in the latest commit. Marking as resolved.